AtomicsIsLockFree should not be eliminated in DCE in JavaScriptCore

hyjorc1 opened the pull request:
#### 561a9d8fc5e3e6526deb2b77923ed72786bbae99
<pre>
AtomicsIsLockFree should not be eliminated in DCE in JavaScriptCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=284245">https://bugs.webkit.org/show_bug.cgi?id=284245</a>
<a href="https://rdar.apple.com/141233504">rdar://141233504</a>

Reviewed by Keith Miller and Yusuke Suzuki.

AtomicsIsLockFree should be eliminated only if it has no side effects.
Here we do the elimination when its argument is not int32 use.

* JSTests/stress/atomics-is-lock-free.js: Added.
(f2):
(f1):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGNodeType.h:

Canonical link: <a href="https://commits.webkit.org/287673@main">https://commits.webkit.org/287673@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0dd15f76d83996bf485f76c5def26436e1a1e0a6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80407 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59413 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34026 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84928 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31389 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68475 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7716 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62835 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20644 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83476 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52920 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73206 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43139 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50231 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27360 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29848 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73387 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71371 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27880 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86362 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79466 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7632 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5396 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71119 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7807 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69043 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70360 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14366 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13303 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101873 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12449 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7594 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24822 "Found 2 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7433 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10953 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9239 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -941,6 +941,11 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
+        if (value.isNumber()) {
+            node->clearFlags(NodeMustGenerate);
+            m_state.setShouldTryConstantFolding(true);
+        }
```
I think this should be 
```
else {
    node->clearFlags(NodeMustGenerate);
    if (JSValue value = m_state.forNode(m_graph.child(node, 0)).value())
        m_state.setShouldTryConstantFolding(true);
}
```
We want to allow the node to be cleared if it has no side effects even if we can't constant fold it. We also need to add a constant folding rule since I don't see one. Or you could just not add the constant folding part.

Constellation commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -941,6 +941,11 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
+            // FIXME: Non-Int32Use is really rare. Ideally we should just treat
+            // this as a DFG node with child1 in Int32Use directly in DFGByteCodeParser.
+            node->clearFlags(NodeMustGenerate);
+        }
```
We need to do clearing in the fixup phase instead.

kmiller68 replied:
I think either/both work? If Fixup sets the use kind AI will see the proof it here and remove the must generate?

Constellation commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -1523,12 +1523,19 @@ class FixupPhase : public Phase {
+            Edge child1 = m_graph.child(node, 0);
+            if (child1->shouldSpeculateInt32())
                 fixIntOrBooleanEdge(m_graph.child(node, 0));
+            if (child1.useKind() != Int32Use) {
```
It is opposite. So, let's do in this way,

```
if (child1->shouldSpeculateInt32()) {
    fixIntOrBooleanEdge(m_graph.child(node, 0));
    node->clearFlags(NodeMustGenerate);
}
```

I think, if node is already handling UntypedUse, probably we do not need FIXME.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -941,6 +942,8 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
         if (m_graph.child(node, 0).useKind() != Int32Use)
             clobberWorld();
+        else
+            node->clearFlags(NodeMustGenerate);
```
We don't need this since it's not even based on the proven type it's not going to do anything.

kmiller68 reviewed (approved):
r=me.

hyjorc1 committed:
AtomicsIsLockFree should not be eliminated in DCE in JavaScriptCore
https://bugs.webkit.org/show_bug.cgi?id=284245
rdar://141233504

Reviewed by Keith Miller and Yusuke Suzuki.

AtomicsIsLockFree should be eliminated only if it has no side effects.
Here we do the elimination when its argument is not int32 use.

* JSTests/stress/atomics-is-lock-free.js: Added.
(f2):
(f1):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGNodeType.h:

Canonical link: https://commits.webkit.org/287673@main
